waiter/src/waiter/service_description.clj

```diff
@@ -747,12 +754,37 @@
                           (merge (error-message-map-fn passthrough-headers waiter-headers) (dissoc ex-data :message))
                           (:throwable &throw-context))))))))
 
+(defn retrieve-fallback-token
+  "Computes the most recently modified token from the token->token-data map."
+  [token->token-data]
+  (first (apply max-key (fn [[_ token-data]] (get-in token-data ["previous" "last-update-time"] 0)) token->token-data)))
```

Comment:
Supporting fallbacks for multiple tokens has complications:

Case 1:
Let's assume we have a service that uses token A and B.  User updates A and B, in arbitrary order.
It seems arbitrary for us to explore the history of A, and not the history of B, or vice versa.

Case 2:
Let's assume user updates A to A'.
User sees that they didn't intend to update A, but really intended to update B.
User updates B to B', and then reverts A' to A.
When we find that [A B'] doesn't exist, we'll never explore for the running service [A B], instead looking for A'

Should we consider searching the space of all possible versions for any given time for running services?  Or should we consider not supporting multiple tokens for fallback?

Reply:
> Should we consider searching the space of all possible versions for any given time for running services?

If we do, from your example, we would end up finding in the history in the following order using the token snapshots based on change timestamps: [A B'], [A' B'], [A' B], [A B]. So we will eventually end up finding the snapshot [A B] and its running service. We are bounded to `num-tokens` * `history-length` searches. We could even add a configuration to limit the lookup length. Also, if the user knows how the tokens will be looked up, they can change their behavior and decide to update A' to A before updating B to B'.

> Or should we consider not supporting multiple tokens for fallback?

That is one approach, but given that the other option is not that much harder, not sure why we would want to limit the search space.